src/dnf-plugins/product-id/product-id.c

```diff
@@ -0,0 +1,640 @@
+                        g_free(abs_file_name);
+                    }
+                    g_free(product_id);
+                }
+            } else if (errno != 0 && errno != ENODATA && errno != EEXIST) {
+                error("Unable to read content of %s directory, %d, %s", PRODUCT_CERT_DIR, errno, strerror(errno));
+            }
+        } while (file_name != NULL);
+        g_dir_close(productDir);
+    } else {
+        printError("Unable to open directory with product certificates", tmp_err);
+    }
+    return 0;
+}
+
+/**
+ * Returns string representation of pluginHookdId. There is no need to free string, because it is statically
+ * allocated.
+ *
+ * @param id Id of plugin hook
+ * @return String representing hook
+ */
+gchar *strHookId(PluginHookId id) {
+    switch(id) {
+        case PLUGIN_HOOK_ID_CONTEXT_PRE_CONF:
+            return "CONTEXT_PRE_CONF";
+        case PLUGIN_HOOK_ID_CONTEXT_CONF:
+            return "CONTEXT_CONF";
+        case PLUGIN_HOOK_ID_CONTEXT_PRE_TRANSACTION:
+            return "PRE_TRANSACTION";
+        case PLUGIN_HOOK_ID_CONTEXT_TRANSACTION:
+            return "CONTEXT_TRANSACTION";
+        case PLUGIN_HOOK_ID_CONTEXT_PRE_REPOS_RELOAD:
+            return "CONTEXT_PRE_REPOS_RELOAD";
+        default:
+            return "UNKNOWN";
+    }
+}
+
+/**
+ * Callback function. This method is executed for every libdnf hook. This callback
+ * is called several times during transaction, but we are interested only in one situation.
+ *
+ * @param handle Pointer on structure with data specific for this plugin
+ * @param id Id of hook (moment of transaction, when this callback is called)
+ * @param hookData
+ * @param error
+ * @return
+ */
+int pluginHook(PluginHandle *handle, PluginHookId id, void *hookData, PluginHookError *error) {
+    // We do not need this for anything
+    (void)hookData;
+    (void)error;
+
+    if (!handle) {
+        // We must have failed to allocate our handle during init; don't do anything.
+        return 0;
+    }
+
+    debug("%s v%s, running hook_id: %s on DNF version %d",
+            pinfo.name, pinfo.version, strHookId(id), handle->version);
+
+    if (id == PLUGIN_HOOK_ID_CONTEXT_TRANSACTION) {
+        // Get DNF context
+        DnfContext *dnfContext = handle->initData;
+        // List of all repositories
+        GPtrArray *repos = dnf_context_get_repos(dnfContext);
+        // When there are no repositories, then we can't do anything
+        if(repos == NULL) {
+            return 1;
+        }
+        // List of enabled repositories
+        GPtrArray *enabledRepos = g_ptr_array_sized_new(repos->len);
+        // Enabled repositories with product id certificate
+        GPtrArray *repoAndProductIds = g_ptr_array_sized_new(repos->len);
+        // Enabled repositories with prouctid cert that are actively used
+        GPtrArray *activeRepoAndProductIds = g_ptr_array_sized_new(repos->len);
+
+        ProductDb *productDb = initProductDb();
```

Comment:
It doesn't look like we are actually reading the existing product DB at any point.  I don't know if that's a mistake or intentional.  FWIF, the python plugin reads the DB before it starts running

Reply:
I don't think it is necessary to read product db till we support cache-only mode. I will add TODO: comment. Thoughts?